sdk/core/azure-core/src/main/java/com/azure/core/util/BinaryData.java

```diff
@@ -125,128 +144,147 @@ public static BinaryData fromStream(InputStream inputStream) {
      * @return A {@link BinaryData} representing the {@link String}.
+     * @throws NullPointerException If {@code data} is null.
      */
     public static BinaryData fromString(String data) {
-        if (CoreUtils.isNullOrEmpty(data)) {
-            return EMPTY_DATA;
-        }
-
-        return new BinaryData(data.getBytes(StandardCharsets.UTF_8));
+        Objects.requireNonNull(data, "'data' cannot be null.");
+        return new BinaryData(new StringContent(data));
     }
 
     /**
      * Creates an instance of {@link BinaryData} from the given byte array.
      * <p>
-     * If the byte array is null or zero length an empty {@link BinaryData} will be returned.
+     * If the byte array is null or zero length an empty {@link BinaryData} will be returned. Note that the input
+     * byte array is used as a reference by this instance of {@link BinaryData} and any changes to the byte array
+     * outside of this instance will result in the contents of this BinaryData instance to be updated as well. To
+     * safely update the byte array, it is recommended to make a copy of the contents first.
+     * </p>
      *
      * <p><strong>Create an instance from a byte array</strong></p>
      *
      * {@codesnippet com.azure.core.util.BinaryData.fromBytes#byte}
      *
      * @param data The byte array that {@link BinaryData} will represent.
      * @return A {@link BinaryData} representing the byte array.
+     * @throws NullPointerException If {@code data} is null.
      */
     public static BinaryData fromBytes(byte[] data) {
-        if (Objects.isNull(data) || data.length == 0) {
-            return EMPTY_DATA;
-        }
-
-        return new BinaryData(Arrays.copyOf(data, data.length));
+        Objects.requireNonNull(data, "'data' cannot be null.");
+        return new BinaryData(new ByteArrayContent(data));
     }
 
     /**
      * Creates an instance of {@link BinaryData} by serializing the {@link Object} using the default {@link
      * JsonSerializer}.
-     * <p>
-     * If {@code data} is null an empty {@link BinaryData} will be returned.
+     *
      * <p>
      * <b>Note:</b> This method first looks for a {@link JsonSerializerProvider} implementation on the classpath. If no
      * implementation is found, a default Jackson-based implementation will be used to serialize the object.
-     *
+     *</p>
      * <p><strong>Creating an instance from an Object</strong></p>
      *
      * {@codesnippet com.azure.core.util.BinaryData.fromObject#Object}
      *
      * @param data The object that will be JSON serialized that {@link BinaryData} will represent.
      * @return A {@link BinaryData} representing the JSON serialized object.
+     * @throws NullPointerException If {@code data} is null.
      * @see JsonSerializer
      */
     public static BinaryData fromObject(Object data) {
+        Objects.requireNonNull(data, "'data' cannot be null.");
```

Comment:
Should allow null data here, the underlying serializer should determine whether this is allowed. For example, Jackson can turn `null` into JSON-null.